<div style="border:solid black 2px; padding: 20px">

# Reviewer's feedback v.1

**Halo Teguh**
    
**Greetings from Chamdani**

Terima kasih telah mengirimkan project kamu dengan ini kamu sudah memulai sebuah langkah yang luar biasa saat ini. Disini saya akan mereview proyek kamu ya.

Saya akan memberikan beberapa komentar dan feedback seperti dibawah ini, **mohon jangan dipindah, dirubah, maupun dihapus ya :).**
    
> Mohon diperhatikan bahwa apabila ada temuan atau kesalahan yang sama namun tidak ada komentar perbaikan disana, itu berarti kamu masih tetap harus memperbaikinya yaa.

Komentar yang saya berikan akan muncul dalam warna hijau, kuning, atau pun merah seperti ini:

</div>

<div class="alert alert-success">
<b>Chamdani's comment v.*</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

<div class="alert alert-warning">
<b>Chamdani's comment v.*</b> <a class="tocSkip"></a>

Ada beberapa catatan.

</div>

<div class="alert alert-danger">
<b>Chamdani's comment v.*</b> <a class="tocSkip"></a>

Perlu beberapa perbaikan.

</div>

<div class="alert alert-info">
<b>Chamdani's info v.1</b> <a class="tocSkip"></a>

Catatan umum:
* Fungsi `print` tidak diperlukan untuk mencetak suatu nilai
* Lebih baik tulis kode pada bagian akhir cell code untuk menampilkan output tanpa `print`
* Perintah soal yang ada dalam tanda kurung siku `[...]` sebaiknya dihilangkan
* Sangat disarankan untuk menggunakan cell markdown untuk memberikan deskripsi temuan atau kesimpulan.
* Tidak disarankan menggunakan cell code untuk memberikan penjelasan / deskripsi.
* Untuk menghindari galat pada notebook project kamu. Batasi jumlah output yang dikeluarkan setiap kode dengan batas max 10 baris data.

</div>

<div style="border:solid gray 5px; padding: 20px" class="alert alert-warning">

**INGATLAH** 
* Project kamu tidak akan bisa diterima bila kamu masih memiliki feedback berwarna merah. 
* Kamu hanya memiliki 3 kali kesempatan iterasi submission disetiap projectnya 
* Jika merasa kesulitan kamu dapat menuliskan responmu atas feedback saya menggunakan blok markdown warna biru dan kamu bisa mengirimkan project pada submission selanjutnya untuk mendapatkan jawaban / respon dari saya segera yaa

Kamu dapat menjawab saya dengan menggunakan blok berwarna biru seperti ini

<div class="alert alert-info">
<b>Student answer</b> <a class="tocSkip"></a>

Siap kak.

</div>

<div style="border:solid black 2px; padding: 20px" class="alert alert-success">
    
# General feedback v.1 *Project Accepted*

Terima kasih banyak telah mengirimkan proyek Anda! 
<br>
    
Secara keseluruhan, proyek Anda **SANGAT** mengesankan. Semoga apa yang kamu pelajari dalam project ini dapat membantu meningkatkan kemampuanmu. Pertahankan apa yang sudah bagus, dan tingkatkan apa yang menurutmu masih perlu ditingkatkan. Semoga berhasil pada project selanjutnya yaa :)

**TETAP SEMANGAT!!**
    
</div>

# Pendahuluan

Project ini dilakukan untuk membangun model yang bisa menganalisis perilaku konsumen dan merekomendasikan salah satu dari kedua paket terbaru Megaline: Smart atau Ultra.

Target dari dataset yang kita miliki yaitu bertipe kategorik `1` dan `0`:

**Target :**

0 : Pengguna tidak berlangganan paket ultra

1 : Pengguna telah berlangganan paket ultra


Dari data yang dimiliki kita memiliki 4 atribut (features) dan 1 target, untuk itu kita akan membangun model dengan metode klasifikasi yaitu algoritma *Decision Tree*, *Random Forest*, *Logistic Regression*. Dari ketiga algoritma tersebut kita akan membandingkan algoritma mana yang menghasilkan performa yang terbaik.

## Informasi Atribut Data


Setiap observasi dalam dataset yang kita miliki memuat informasi perilaku bulanan tentang satu pengguna. Informasi tersebut mencakup: 

| Atribut | Tipe Data | Deskripsi |
| --- | --- | --- |
| calls | Float | Jumlah total panggilan |
| minutes | Float | Total durasi panggilan dalam satuan menit | 
| messages | Float | Jumlah total pesan teks |
| mb_used | Float | Traffic penggunaan internet dalam satuan MB |
| is_ultra | Integer | Apakah sedang menggunakan paket ultra ? (1: Ya, 0: Tidak) |

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Hebat, Kerja bagus!

</div>

## Memahami Data

### Mengimport Library

In [2]:
# library yang dibutuhkan
import pandas as pd
import numpy as np

from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import train_test_split
# metriks
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, precision_score, recall_score, f1_score
# klasifikasi
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Hebat, Kerja bagus!

</div>

### Memuat dan Pemahaman Data

In [3]:
# memuat dataset
try:
    df = pd.read_csv('/datasets/users_behavior.csv')
except:
    df = pd.read_csv('users_behavior.csv')
# menampilkan data
df.head()

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

In [4]:
# informasi dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


In [5]:
pd.set_option('display.max_colwidth', 0)
# Melihat Deskripsi pengguna kedua paket
display(df.query('is_ultra == 1').describe(include='all').transpose().style.set_caption('Deskripsi Pengguna Paket Ultra'),
       df.query('is_ultra == 0').describe(include='all').transpose().style.set_caption('Deskripsi Pengguna Paket Smart'))

In [6]:
# pengaturan ukuran tabel
pd.set_option('display.max_colwidth', 0)

# membuat tabel dengan unique value dan jumlahnya menggunakan loop
list_item = []
for col in df.columns:
    list_item.append([col, df[col].nunique(), df[col].unique()])
    
unique_table = pd.DataFrame(data=list_item, columns=['column','number_of_unique','unique_item'])
unique_table

column  number_of_unique  \
0  calls     184                
1  minutes   3144               
2  messages  180                
3  mb_used   3203               
4  is_ultra  2                  

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               unique_item  
0  [40.0, 85.0, 77.0, 106.0, 66.0, 58.0, 57.0, 15.0, 7.0, 90.0, 82.0, 45.0, 51.0, 56.0, 108.0, 6.0, 2.0, 26.0, 79.0, 49.0, 93.0, 48.0, 11.0, 53.0, 81.0, 154.0, 37.0, 50.0, 41.0, 10.0, 71.0, 65.0, 110.0, 120.0, 76.0, 64.0, 23.0, 34.0, 98.0, 35.0, 5.0, 70.0, 124.0, 129.0, 67.0, 0.0, 13.0, 68.0, 91.0, 121.0, 114.0, 125.0, 80.0, 33.0, 138.0, 84.0, 78.0, 69.0, 63.0, 72.0, 73.0, 1.0, 43.0, 118.0, 74.0, 83.0, 141.0, 117.0, 54.0, 101.0, 29.0, 3.0, 107.0, 55.0, 47.0, 158.0, 87.0, 28.0, 59.0, 52.0, 44.0, 17.0, 111.0, 109.0, 14.0, 92.0, 94.0, 46.0, 133.0, 75.0, 38.0, 60.0, 100.0, 31.0, 61.0, 89.0, 27.0, 196.0, 24.0, 99.0, ...]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             
1  [311.8999999999999, 516.75, 467.6600000000001, 745.53, 418.74, 344.56, 431.6400000000001, 132.4, 43.39, 665.4100000000001, 560.5099999999998, 344.32, 437.1299999999999, 433.0699999999999, 587.9000000000001, 22.13, 18.73, 163.62000000000003, 532.62, 341.6699999999999, 691.37, 411.46, 351.6, 96.04999999999998, 360.3, 356.22, 624.71, 1078.64, 286.67999999999995, 358.01000000000005, 276.80999999999995, 77.09, 530.13, 482.35, 776.8499999999999, 814.8500000000004, 543.1799999999998, 422.6899999999999, 146.18, 519.48, 183.28, 85.31, 392.04, 323.98999999999995, 665.5599999999998, 235.58, 535.9099999999999, 327.45000000000005, 49.95, 474.9200000000001, 532.1000000000001, 842.3999999999999, 929.2300000000002, 421.27000000000015, 0.0, 106.03, 449.2900000000001, 585.09, 477.5399999999997, 356.3900000000001, 853.01, 823.5799999999999, 622.6600000000001, 814.52999999

In [7]:
# melihat jumlah baris dan kolom
print(f"Baris Data : {df.shape[0]}")
print(f"Kolom Data: {df.shape[1]}")

Baris Data : 3214
Kolom Data: 5


In [8]:
# menghitung duplikat
df.duplicated().sum()

0

In [9]:
# menghitung vissing value disetiap kolom
missing_report = df.isna().sum().to_frame()
missing_report.columns = ['missing_count']
missing_report['percent_of_all'] = (missing_report.missing_count / len(df)).round(2).astype(str) + '%'
missing_report

missing_count percent_of_all
calls     0              0.0%         
minutes   0              0.0%         
messages  0              0.0%         
mb_used   0              0.0%         
is_ultra  0              0.0%

**Kesimpulan**

1. Kita memiliki `3124` baris data dan `5` kolom.
2. Dari dataset kita dapat melihat bahwa pengguna paket ultra sebanyak `985` dan pengguna paket smart sebanyak `2229`.
3. Dataset yang kita punya untuk features semuanya merupakan variabel kuantitatif, sedangkan untuk targetnya (`is_ultra`) merupakan variabel kategorik dimana bernilai `1` dan `0`. 
4. Kita tidak mempunyai variabel seperti `id` ataupun lainnya untuk melihat apakah terdapat duplikat eksplisit jadi untuk saat ini bisa dipastikan bahwa tabel kita tidak mempunyai duplikat sesuai dengan pemeriksaan diatas.
5. Pada dataset juga tidak terdapat missing value.
6. Beberapa kolom kita akan mengubah tipe datanya menjadi `integer` ataupun `boolean`, yaitu:
    * `calls` - karena ini menghitung total jumlah panggilan pengguna jadi akan lebih baik jika tipe datanya `integer`.
    * `messages` - kolom ini juga menghitung total pesan teks pengguna jadi akan lebih baik jika tipe datanya `integer`.
    * `is_ultra` - kolom ini merupakana variabel kategorik yang bernilai `1` dan `0`.

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Hebat, Kerja bagus!

</div>

### Mengubah Tipe Data

In [10]:
#mengubah tipe data pada kolom yang dibutuhkan
columns = ['calls','messages','is_ultra']

for col in columns:
    if col != 'is_ultra':
        df[col] = df[col].astype('int')
    else:
        df[col] = df[col].astype('bool')

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   int64  
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   int64  
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   bool   
dtypes: bool(1), float64(2), int64(2)
memory usage: 103.7 KB


<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Hebat, Kerja bagus!

</div>

## Membagi Data Training, Data Validation, dan Data Testing

Dalam membagi data yang kita punya, aturan untuk membagi data validation dan data testing yaitu mempunyai ukuran yang sama. Jadi kita akan membagi data dengan komposisi seperti ini:
- Data Training - 60%
- Data Validation - 20%
- Data Testing - 20%

In [12]:
# Data Training dengan komposisi 60%
df_train, df_validation = train_test_split(df, train_size= .60, random_state=45)

In [13]:
# Data Validation komposisi 20% dan Data Testing 20%
df_valid, df_test = train_test_split(df_validation, test_size= .50, random_state=45)

In [14]:
train_comp = len(df_train) / len(df)
print(f"Komposisi Data Training : {train_comp:.0%}")

Komposisi Data Training : 60%


In [15]:
valid_comp = len(df_valid) / len(df)
print(f"Komposisi Data Validation : {valid_comp:.0%}")

Komposisi Data Validation : 20%


In [16]:
test_comp = len(df_test) / len(df)
print(f"Komposisi Data Testing : {test_comp:.0%}")

Komposisi Data Testing : 20%


<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Hebat, Kerja bagus!

</div>

## Membagi Data Features dan Target

Setelah kita memahami dataset kita. Kita akan membagi dataset tersebut menjadi 2 bagian:
- **features** - data yang merupakan representasi struktural dari data yang digunakan untuk melatih model mesin, melakukan prediksi, atau melakukan tugas analisis data.
- **target** - merupakan komponen yang menyimpan informasi atau label yang ingin diprediksi atau dimodelkan oleh algoritma machine learning.

**Data Features**

In [17]:
# fitur untuk data training
features_train = df_train.drop(['is_ultra'], axis=1)
# fitur untuk data validation
features_valid = df_valid.drop(['is_ultra'], axis=1)
# fitur untuk data testing
features_test = df_test.drop(['is_ultra'], axis=1)

**Data Target**

In [18]:
# target untuk data_train
target_train = df_train.is_ultra
# target untuk data validation
target_valid = df_valid.is_ultra
# target untuk data testing
target_test = df_test.is_ultra

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Hebat, Kerja bagus!

</div>

## Melatih Model 

### Algoritma Decision Tree

In [19]:
DTClass = DecisionTreeClassifier(random_state = 45)
DTClass.fit(features_train, target_train)

pred_train = DTClass.predict(features_train)
accuracy_train = accuracy_score(target_train, pred_train)
print(f"Akurasi Data Training: {accuracy_train}")
#
pred_valid = DTClass.predict(features_valid)
accuracy_valid = accuracy_score(target_valid, pred_valid)
print(f"Akurasi Data Valid: {accuracy_valid}")

Akurasi Data Training: 1.0
Akurasi Data Valid: 0.7231726283048211


Model kita sepertinya masih overfitting, dimana akurasi pada data validationnya hanya mencapai 73%. Sedangkan akurasi minimalnya yaitu 75%, untuk itu kita akan melakukan tuning hyper-parameter guna meningkatkan akurasi dari model kita.

In [20]:
# variabel untuk menyimpan nilai depth, sample_leaf, dan accuracy terbaik
best_depth = 0
best_min_samples_leaf = 0
best_accuracy = 0

In [21]:
# looping untuk mencari kedalaman dengan akurasi terbaik dari rentang 1:20
for depth in range(1, 20):
    DTClass = DecisionTreeClassifier(random_state = 45, max_depth = depth)
    DTClass.fit(features_train, target_train)
    #
    score = DTClass.score(features_valid, target_valid)
    if score > best_accuracy:
        best_accuracy = score
        best_depth = depth
#     print(f"Max Depth : {depth} Accuracy: {score}")

In [22]:
print(f"Max Depth : {best_depth}\nAccuracy: {best_accuracy}")

Max Depth : 10
Accuracy: 0.8009331259720062


In [23]:
#looping untuk mencari nilai min_samples_leaf
for leaf in range(1, 20):
    DTClass = DecisionTreeClassifier(random_state = 45, max_depth = best_depth, min_samples_leaf = leaf)
    DTClass.fit(features_train, target_train)
    score = DTClass.score(features_valid, target_valid)
    if score >= best_accuracy:
        best_accuracy = score
        best_min_samples_leaf = leaf

In [24]:
print(f"Min Sample Leaf : {best_min_samples_leaf}\nMax_depth: {best_depth}\nAccuracy: {best_accuracy}")

Min Sample Leaf : 1
Max_depth: 10
Accuracy: 0.8009331259720062


Sepertinya data validation sudah mencapai akurasi terbaiknya yang bernilai 80% dan dapat dilihat dari sini bahwa dengan mengatur hyper-parameter min_samples_leaf dari 1 s/d 20, nilai terbaiknya yaitu dengan min_samples_leaf 1 dan hyper-parameter ini nilai defaultnya juga 1. maka kita lanjut untuk sanity check pada data test kita menggunakan hyper parameter dengan max_depth = 10

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Hebat, Kerja bagus!

</div>

In [25]:
# menguji data testing dengan hyper-parameter terbaik
DTClass = DecisionTreeClassifier(random_state = 45, max_depth = best_depth)
DTClass.fit(features_train, target_train)

# data validation
pred_valid = DTClass.predict(features_valid)
accuracy_valid = accuracy_score(target_valid, pred_valid)
print(f"Akurasi Data Validation: {accuracy_valid}")
# data testing
pred_test = DTClass.predict(features_test)
accuracy_test = accuracy_score(target_test, pred_test)
print(f"Akurasi Data Testing: {accuracy_test}")

Akurasi Data Validation: 0.8009331259720062
Akurasi Data Testing: 0.7869362363919129


In [26]:
# nilai precision
precision = precision_score(target_test, pred_test)
precision

0.7191780821917808

In [27]:
# nilai recall
recall = recall_score(target_test, pred_test)
recall

0.5223880597014925

**Kesimpulan** 

Dengan algoritma Decision Tree dan dengan mengatur hyper-parameter `max_depth` menjadi 10, kita mendapatkan akurasi data validation sekitar 0.80. Dan saat diuji kedata testing kita mendapatkan hasil berikut:

| Akurasi | Precision | Recall |
| --- | --- | --- |
| 0.78 | 0.71 | 0.52 |

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Hebat, Kerja bagus!

</div>

### Algoritma Random Forest

In [28]:
# melatih model tanpa tanpa merubah parameter `n_estimators`
RFClass = RandomForestClassifier(random_state = 45)
RFClass.fit(features_train, target_train) # pelatihan model

pred_train = RFClass.predict(features_train)
accuracy_train = accuracy_score(target_train, pred_train)
print(f"Akurasi Data Training: {accuracy_train}")
#
pred_valid = RFClass.predict(features_valid)
accuracy_valid = accuracy_score(target_valid, pred_valid)
print(f"Akurasi Data Validation: {accuracy_valid}")

Akurasi Data Training: 1.0
Akurasi Data Validation: 0.7869362363919129


In [29]:
best_score = 0
best_est = 0

In [30]:
# mencari akurasi terbaik dgn mencari n_estimators
for est in range(1, 22):
    RFClass = RandomForestClassifier(random_state = 45, n_estimators = est)
    RFClass.fit(features_train, target_train)
    pred_valid = RFClass.predict(features_valid)
    accuracy_valid = accuracy_score(target_valid, pred_valid)
    if accuracy_valid > best_score:
        best_score = accuracy_valid # simpan skor accuracy terbaik pada validation set
        best_est = est

In [31]:
print(f"best akurasi : {best_score}, best_est: {best_est}")

best akurasi : 0.8040435458786936, best_est: 20


In [32]:
# mencoba mengatur hyper-parameter bootstrap = False
RFClass = RandomForestClassifier(random_state = 45, n_estimators = best_est, bootstrap = False)
RFClass.fit(features_train, target_train)
pred_valid = RFClass.predict(features_valid)
accuracy_valid = accuracy_score(target_valid, pred_valid)
print(f"bootstrap = False : {accuracy_valid}")

bootstrap = False : 0.7511664074650077


Akurasi yang didapatkan dapat meningkat 2% dengan mengatur hyper-parameter `n_estimators` menjadi 20 dan tidak ada hasil yang lebih baik dari pada itu, jadi kita akan mengatur hyper-parameter hanya untuk `n_estimators` = 20 untuk menguji data testing.

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Hebat, Kerja bagus!

</div>

In [33]:
# menguji data testing
RFClassClass = RandomForestClassifier(random_state = 45, n_estimators = 20)
RFClass.fit(features_train, target_train)
pred_test = RFClass.predict(features_test)
accuracy_test = accuracy_score(target_test, pred_test)

In [34]:
print(f"Akurasi Data Testing: {accuracy_test}")

Akurasi Data Testing: 0.7931570762052877


In [35]:
# nilai precision
precision = precision_score(target_test, pred_test)
precision

0.6954022988505747

In [36]:
# nilai recall
recall = recall_score(target_test, pred_test)
recall

0.6019900497512438

**Kesimpulan**

Dengan menggunakan algoritma random forest dan mengatur hyper-parameter `n_estimators` menjadi 20, kita mendapatkan akurasi tertinggi untuk data validation yaitu 80%, dan metriks untuk data testing yaitu:

| Akurasi | Precision | Recall |
| --- | --- | --- |
| 0.79 | 0.69 | 0.60 |

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Hebat, Kerja bagus!

</div>

### Algoritma Regresi Logistik

In [37]:
logreg = LogisticRegression(random_state = 45, solver='liblinear')
logreg.fit(features_train, target_train)

pred_train = logreg.predict(features_train)
accuracy_train = accuracy_score(target_train, pred_train)
print(f"Akurasi Data Training: {accuracy_train}")
#
pred_valid = logreg.predict(features_valid)
accuracy_valid = accuracy_score(target_valid, pred_valid)
print(f"Akurasi Data Validation: {accuracy_valid}")

Akurasi Data Training: 0.7090248962655602
Akurasi Data Validation: 0.6702954898911353


In [38]:
param_grid = {
    'C': [0.001, 0.01, 0.1, 1, 10, 10.5, 100],
    'penalty': ['l1', 'l2'],
    'max_iter' : [101, 126, 152, 182, 216, 241, 256]
}

grid_search = GridSearchCV(LogisticRegression(random_state=45, solver='liblinear'), param_grid, cv=10)
grid_search.fit(features_train, target_train)

best_params = grid_search.best_params_
best_model = grid_search.best_estimator_

In [40]:
best_model.fit(features_train, target_train)

test_valid = best_model.predict(features_valid)
accuracy_valid = accuracy_score(target_valid, test_valid)
accuracy_valid

0.7247278382581649

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Hebat, Kerja bagus!

</div>

In [39]:
logreg = LogisticRegression(C=10, max_iter=101, penalty='l1', random_state= 45, solver='liblinear')
logreg.fit(features_train, target_train)

pred_train = logreg.predict(features_train)
accuracy_train = accuracy_score(target_train, pred_train)
print(f"Akurasi Data Training: {accuracy_train}")
#
pred_valid = logreg.predict(features_valid)
accuracy_valid = accuracy_score(target_valid, pred_valid)
print(f"Akurasi Data Validation: {accuracy_valid}")

Akurasi Data Training: 0.7588174273858921
Akurasi Data Validation: 0.7247278382581649


In [40]:
# melihat akurasi pada data testing
pred_test = logreg.predict(features_test)
accuracy_test = accuracy_score(target_test, pred_test)
print(f"Akurasi Data Testing: {accuracy_test}")

Akurasi Data Testing: 0.7418351477449455


In [41]:
# nilai precision
precision = precision_score(target_test, pred_test)
precision

0.7536231884057971

In [42]:
# nilai recall
recall = recall_score(target_test, pred_test)
recall

0.25870646766169153

**Kesimpulan**

Dengan algoritma Regresi Logistik dengan parameter `C=1, max_iter=101, penalty='l1', random_state=45, solver='liblinear'` didapatkan hasil akurasi yang maksimal, namun pada data yang kita miliki angka akurasi dari algoritma ini tidak mencapai angka 75%, hasil untuk data testingnya seperti berikut ini:

| Akurasi | Precision | Recall |
| --- | --- | --- |
| 0.74 | 0.75 | 0.25 |

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Hebat, Kerja bagus!

</div>

## Kesimpulan Umum

Dari peneletian ini dengan melakukan perbandingan menggunakan tiga(3) algoritma, *Decision Tree*, *Random Forest* , *Logistic Regression* hasil akurasi tertinggi yaitu dengan menggunakan algoritma *Random Forest* dengan nilai akurasi 79%, kemudian algoritma *Decision Tree* dengan akurasi 78% dan algoritma *Logistic Regression* yang akurasinya paling rendah yaitu 74%.

<div class="alert alert-success">
<b>Chamdani's comment v.1</b> <a class="tocSkip"></a>

Hebat, Kerja bagus!

</div>